File: lib/src/main/java/com/orientsec/easysocket/Options.java
package com.orientsec.easysocket;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.orientsec.easysocket.client.ClientInitializer;
import com.orientsec.easysocket.push.PushManager;
import com.orientsec.easysocket.request.Decoder;
import com.orientsec.easysocket.request.Request;
import com.orientsec.easysocket.session.SessionInitializer;
import com.orientsec.easysocket.utils.Executors;

import java.util.List;
import java.util.concurrent.Executor;

import javax.net.SocketFactory;

/**
 * Represents the configuration options for the EasySocket library.
 * This class provides various settings for socket connections, including
 * connection timeouts, heartbeat configurations, reconnection policies, and
 * thread executors for managing tasks.
 */
public class Options {

    /**
     * Indicates whether the application is in debug mode.
     */
    private final boolean debuggable;
    /**
     * The minimum log level for outputting logs.
     */
    private final int minLogLevel;

    /**
     * The name of the socket client.
     */
    private final String name;

    /**
     * Provider for the heartbeat decoder.
     */
    private final Provider<Decoder<Boolean>> pulseDecoderProvider;

    /**
     * Provider for the heartbeat request.
     */
    private final Provider<Request<Boolean>> pulseRequestProvider;

    /**
     * Provider for the socket factory.
     */
    private final Provider<SocketFactory> socketFactoryProvider;

    /**
     * Provider for the data protocol parser.
     */
    private final Provider<HeadParser> headParserProvider;

    /**
     * Provider for the push message handler.
     */
    private final Provider<PushManager<?, ?>> pushManagerProvider;
    /**
     * Provider for the socket client initializer.
     */
    private final Provider<ClientInitializer> clientInitializerProvider;

    /**
     * Provider for the session initializer.
     */
    private final Provider<SessionInitializer> sessionInitializerProvider;

    /**
     * Executor for handling callback tasks, such as connection state changes
     * and request responses.
     */
    private final Executor callbackExecutor;

    /**
     * Executor for managing connection tasks, such as starting and stopping connections.
     */
    private final Executor connectExecutor;

    /**
     * Executor for encoding and decoding tasks.
     */
    private final Executor codecExecutor;

    /**
     * Executor for handling write operations.
     */
    private final Executor writeExecutor;

    /**
     * A list of server addresses to which the socket client can connect.
     * This list must not be empty.
     */
    private final List<Address> addressList;

    /**
     * The maximum size of data (in KB) that can be read to prevent memory overflow.
     */
    private final int maxReadSizeInKB;

    /**
     * The timeout duration (in seconds) for requests.
     */
    private final int requestTimeOutInMills;

    /**
     * The timeout duration (in seconds) for establishing a connection.
     */
    private final int connectTimeOutInMills;

    /**
     * The timeout duration (in seconds) for io read and write.
     */
    private final int sslTimeOutInMills;

    /**
     * The frequency (in seconds) of heartbeat messages.
     */
    private final int pulseIntervalInSec;

    /**
     * The number of consecutive heartbeat failures allowed before considering
     * the connection lost.
     */
    private final int pulseMaxLostTimes;

    /**
     * The duration (in seconds) the client remains active in the background.
     */
    private final int backgroundActiveDurationInSec;

    /**
     * The reconnection policy to be used when the connection is lost.
     */
    private final ReconnectPolicy reconnectPolicy;

    /**
     * The number of retry attempts for reconnection.
     */
    private final int retryTimesPerAddress;

    /**
     * The interval (in milliseconds) between connection attempts.
     */
    private final int connectIntervalInMills;

    /**
     * A tag used to identify connection statistics.
     */
    private final int connectStatsTag;

    /**
     * A tag used to identify read operation statistics.
     */
    private final int readStatsTag;

    /**
     * A tag used to identify write operation statistics.
     */
    private final int writeStatsTag;
    /**
     * The number of retry attempts for each task.
     */
    private final int taskRetryTimes;

    /**
     * Constructs an `Options` instance using the provided builder.
     *
     * @param builder The builder containing the configuration settings.
     */
    private Options(Builder builder) {
        name = builder.name;
        debuggable = builder.debuggable;
        minLogLevel = builder.minLogLevel;
        callbackExecutor = builder.callbackExecutor;
        connectExecutor = builder.connectExecutor;
        codecExecutor = builder.codecExecutor;
        writeExecutor = builder.writeExecutor;
        maxReadSizeInKB = builder.maxReadSizeInKB;
        requestTimeOutInMills = builder.requestTimeOutInMills;
        connectTimeOutInMills = builder.connectTimeOutInMills;
        sslTimeOutInMills = builder.sslTimeOutInMills;
        pulseIntervalInSec = builder.pulseDurationInSec;
        pulseMaxLostTimes = builder.pulseMaxLostTimes;
        backgroundActiveDurationInSec = builder.backgroundActiveDurationInSec;
        reconnectPolicy = builder.reconnectPolicy;
        retryTimesPerAddress = builder.retryTimesPerAddress;
        connectIntervalInMills = builder.connectIntervalInMills;
        headParserProvider = builder.headParserProvider;
        clientInitializerProvider = builder.clientInitializerProvider;
        sessionInitializerProvider = builder.sessionInitializerProvider;
        pulseRequestProvider = builder.pulseRequestProvider;
        pulseDecoderProvider = builder.pulseDecoderProvider;
        pushManagerProvider = builder.pushManagerProvider;
        socketFactoryProvider = builder.socketFactoryProvider;
        connectStatsTag = builder.connectStatsTag;
        readStatsTag = builder.readStatsTag;
        writeStatsTag = builder.writeStatsTag;
        addressList = builder.addressList;
        taskRetryTimes = builder.taskRetryTimes;
    }

    // Getter methods for accessing the configuration options...

    public boolean isDebuggable() {
        return debuggable;
    }

    public int getMinLogLevel() {
        return minLogLevel;
    }

    public String getName() {
        return name;
    }

    @NonNull
    public Executor getCallbackExecutor() {
        return callbackExecutor;
    }

    @NonNull
    public Executor getConnectExecutor() {
        return connectExecutor;
    }

    @NonNull
    public Executor getCodecExecutor() {
        return codecExecutor;
    }

    @NonNull
    public Executor getWriteExecutor() {
        return writeExecutor;
    }

    public int getMaxReadSizeInKB() {
        return maxReadSizeInKB;
    }

    public int getRequestTimeOutInMills() {
        return requestTimeOutInMills;
    }

    public int getPulseIntervalInSec() {
        return pulseIntervalInSec;
    }

    public int getPulseMaxLostTimes() {
        return pulseMaxLostTimes;
    }

    public int getBackgroundActiveDurationInSec() {
        return backgroundActiveDurationInSec;
    }

    @NonNull
    public ReconnectPolicy getReconnectPolicy() {
        return reconnectPolicy;
    }

    public int getConnectTimeOutInMills() {
        return connectTimeOutInMills;
    }

    public int getSslTimeOutInMills() {
        return sslTimeOutInMills;
    }

    public int getRetryTimesPerAddress() {
        return retryTimesPerAddress;
    }

    public int getConnectIntervalInMills() {
        return connectIntervalInMills;
    }

    @Nullable
    public Provider<Request<Boolean>> getPulseRequestProvider() {
        return pulseRequestProvider;
    }

    @Nullable
    public Provider<Decoder<Boolean>> getPulseDecoderProvider() {
        return pulseDecoderProvider;
    }

    @NonNull
    public Provider<SocketFactory> getSocketFactoryProvider() {
        return socketFactoryProvider;
    }

    @NonNull
    public Provider<HeadParser> getHeadParserProvider() {
        return headParserProvider;
    }

    @Nullable
    public Provider<PushManager<?, ?>> getPushManagerProvider() {
        return pushManagerProvider;
    }

    @Nullable
    public Provider<ClientInitializer> getClientInitializerProvider() {
        return clientInitializerProvider;
    }

    @Nullable
    public Provider<SessionInitializer> getSessionInitializerProvider() {
        return sessionInitializerProvider;
    }

    public int getConnectStatsTag() {
        return connectStatsTag;
    }

    public int getReadStatsTag() {
        return readStatsTag;
    }

    public int getWriteStatsTag() {
        return writeStatsTag;
    }

    @Nullable
    public List<Address> getAddressList() {
        return addressList;
    }

    public int getTaskRetryTimes() {
        return taskRetryTimes;
    }

    /**
     * Builder class for constructing `Options` instances.
     * Provides methods for setting various configuration parameters.
     */
    public static final class Builder {
        // The name of the socket client.
        private String name = "";
        // Indicates whether the application is in debug mode.
        private boolean debuggable;
        // Minimum log level for outputting logs.
        private int minLogLevel = Log.INFO;
        // Provider for the heartbeat request.
        private Provider<Request<Boolean>> pulseRequestProvider;
        // Provider for the heartbeat decoder.
        private Provider<Decoder<Boolean>> pulseDecoderProvider;
        // Provider for the socket factory.
        private Provider<SocketFactory> socketFactoryProvider;
        // Provider for the data protocol parser.
        private Provider<HeadParser> headParserProvider;
        // Provider for the push message handler.
        private Provider<PushManager<?, ?>> pushManagerProvider;
        // Provider for the socket client initializer.
        private Provider<ClientInitializer> clientInitializerProvider;
        // Provider for the session initializer.
        private Provider<SessionInitializer> sessionInitializerProvider;
        // Executor for handling callback tasks.
        private Executor callbackExecutor;
        // Executor for managing connection tasks.
        private Executor connectExecutor;
        // Executor for encoding and decoding tasks.
        private Executor codecExecutor;
        // Executor for handling write operations.
        private Executor writeExecutor;
        // A list of server addresses for the socket client.
        private List<Address> addressList;
        // Maximum size of data (in KB) that can be read.
        private int maxReadSizeInKB = 1024;
        // Timeout duration (in milliseconds) for requests.
        private int requestTimeOutInMills = 5000;
        // Timeout duration (in milliseconds) for connections.
        private int connectTimeOutInMills = 5000;
        // Timeout duration (in milliseconds) for io.
        private int sslTimeOutInMills = 5000;
        // Frequency (in seconds) of heartbeat messages.
        private int pulseDurationInSec = 60;
        // Number of consecutive heartbeat failures allowed.
        private int pulseMaxLostTimes = 2;
        // Duration (in seconds) the client remains active in the background.
        private int backgroundActiveDurationInSec = 30;
        // Reconnection policy for lost connections.
        private ReconnectPolicy reconnectPolicy = ReconnectPolicy.ACTIVE;
        // Number of retry attempts for reconnection.
        private int retryTimesPerAddress;
        // Interval (in milliseconds) between connection attempts.
        private int connectIntervalInMills = 3000;
        // Tag for connection statistics.
        private int connectStatsTag = 0x1001;
        // Tag for read operation statistics.
        private int readStatsTag = 0x1002;
        // Tag for write operation statistics.
        private int writeStatsTag = 0x1003;
        // Number of retry attempts for a task
        public int taskRetryTimes = 2;

        /**
         * Default constructor for the `Builder` class.
         */
        public Builder() {
        }

        /**
         * Sets the name of the socket client.
         *
         * @param val The name to set.
         * @return This builder instance for chaining.
         */
        public Builder name(@NonNull String val) {
            name = val;
            return this;
        }

        /**
         * Enables or disables debug mode.
         *
         * @param val `true` to enable debug mode, `false` to disable it.
         * @return This builder instance for chaining.
         */
        public Builder debuggable(boolean val) {
            debuggable = val;
            return this;
        }

        /**
         * Sets the minimum log level for outputting logs.
         *
         * @param val The minimum log level to set.
         * @return This builder instance for chaining.
         * @see android.util.Log  Levels: Log.VERBOSE, Log.DEBUG, Log.INFO, Log.WARN, Log.ERROR
         */
        public Builder minLogLevel(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("Minimum log level must be non-negative.");
            }
            minLogLevel = val;
            return this;
        }

        /**
         * Sets the provider for the heartbeat request.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder pulseRequestProvider(@NonNull Provider<Request<Boolean>> val) {
            pulseRequestProvider = val;
            return this;
        }

        /**
         * Sets the provider for the heartbeat decoder.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder pulseDecoderProvider(@NonNull Provider<Decoder<Boolean>> val) {
            pulseDecoderProvider = val;
            return this;
        }

        /**
         * Sets the provider for the data protocol parser.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder headParserProvider(@NonNull Provider<HeadParser> val) {
            headParserProvider = val;
            return this;
        }

        /**
         * Sets the provider for the push message handler.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder pushManagerProvider(@NonNull Provider<PushManager<?, ?>> val) {
            pushManagerProvider = val;
            return this;
        }

        /**
         * Sets the provider for the socket client initializer.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder clientInitializerProvider(@NonNull Provider<ClientInitializer> val) {
            clientInitializerProvider = val;
            return this;
        }

        /**
         * Sets the provider for the session initializer.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder sessionInitializerProvider(@NonNull Provider<SessionInitializer> val) {
            sessionInitializerProvider = val;
            return this;
        }

        /**
         * Sets the provider for the socket factory.
         *
         * @param val The provider to set.
         * @return This builder instance for chaining.
         */
        public Builder socketFactoryProvider(@NonNull Provider<SocketFactory> val) {
            socketFactoryProvider = val;
            return this;
        }

        /**
         * Sets the executor for handling callback tasks.
         *
         * @param val The executor to set.
         * @return This builder instance for chaining.
         */
        public Builder callbackExecutor(@NonNull Executor val) {
            callbackExecutor = val;
            return this;
        }

        /**
         * Sets the executor for managing connection tasks.
         *
         * @param val The executor to set.
         * @return This builder instance for chaining.
         */
        public Builder connectExecutor(@NonNull Executor val) {
            connectExecutor = val;
            return this;
        }

        /**
         * Sets the executor for encoding and decoding tasks.
         *
         * @param val The executor to set.
         * @return This builder instance for chaining.
         */
        public Builder codecExecutor(@NonNull Executor val) {
            codecExecutor = val;
            return this;
        }

        /**
         * Sets the executor for handling write operations.
         *
         * @param val The executor to set.
         * @return This builder instance for chaining.
         */
        public Builder writeExecutor(@NonNull Executor val) {
            writeExecutor = val;
            return this;
        }

        /**
         * Sets the list of server addresses for the socket client.
         * This list must not be empty; otherwise, an exception will be thrown.
         *
         * @param val The list of addresses to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the address list is empty.
         */
        public Builder addressList(@NonNull List<Address> val) {
            if (val.isEmpty()) {
                throw new IllegalArgumentException("Address list cannot be empty.");
            }
            addressList = val;
            return this;
        }

        /**
         * Sets the maximum size of data (in KB) that can be read.
         *
         * @param val The maximum size to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is not positive.
         */
        public Builder maxReadSizeInKB(int val) {
            if (val <= 0) {
                throw new IllegalArgumentException("Max read data size in KB must be positive.");
            }
            maxReadSizeInKB = val;
            return this;
        }

        /**
         * Sets the timeout duration (in milliseconds) for requests.
         *
         * @param val The timeout duration to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is not positive.
         */
        public Builder requestTimeOutInMills(int val) {
            if (val <= 0) {
                throw new IllegalArgumentException("Request time out must be positive.");
            }
            requestTimeOutInMills = val;
            return this;
        }

        /**
         * Sets the timeout duration (in milliseconds) for connections.
         *
         * @param val The timeout duration to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is not positive.
         */
        public Builder connectTimeOutInMills(int val) {
            if (val <= 0) {
                throw new IllegalArgumentException("Connect time out must be positive.");
            }
            connectTimeOutInMills = val;
            return this;
        }

        /**
         * Sets the timeout duration (in milliseconds) for connections.
         *
         * @param val The timeout duration to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is not positive.
         */
        public Builder sslTimeOutInMills(int val) {
            if (val <= 0) {
                throw new IllegalArgumentException("So time out must be positive.");
            }
            sslTimeOutInMills = val;
            return this;
        }

        /**
         * Sets the frequency (in seconds) of heartbeat messages.
         *
         * @param val The frequency to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is less than 30 seconds.
         */
        public Builder pulseDurationInSec(int val) {
            if (val < 30) {
                throw new IllegalArgumentException("Pulse rate must be at least 30 seconds.");
            }
            pulseDurationInSec = val;
            return this;
        }

        /**
         * Sets the number of consecutive heartbeat failures allowed.
         *
         * @param val The number of failures to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is negative.
         */
        public Builder pulseMaxLostTimes(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("Pulse lost times cannot be negative.");
            }
            pulseMaxLostTimes = val;
            return this;
        }

        /**
         * Sets the duration (in seconds) the client remains active in the background.
         *
         * @param val The duration to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is negative.
         */
        public Builder backgroundActiveDurationInSec(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("Live time must be positive.");
            }
            backgroundActiveDurationInSec = val;
            return this;
        }

        /**
         * Sets the reconnection policy for lost connections.
         *
         * @param val The reconnection policy to set.
         * @return This builder instance for chaining.
         */
        public Builder reconnectPolicy(@NonNull ReconnectPolicy val) {
            reconnectPolicy = val;
            return this;
        }

        /**
         * Sets the number of retry attempts for reconnection.
         *
         * @param val The number of attempts to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is negative.
         */
        public Builder retryTimesPerAddress(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("Retry times cannot be negative.");
            }
            retryTimesPerAddress = val;
            return this;
        }

        /**
         * Sets the interval (in milliseconds) between connection attempts.
         *
         * @param val The interval to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is less than or equal to 1000 milliseconds.
         */
        public Builder connectIntervalInMills(int val) {
            if (val <= 1000) {
                throw new IllegalArgumentException("Connect interval must be greater than" +
                        " 1000 milliseconds.");
            }
            connectIntervalInMills = val;
            return this;
        }

        /**
         * Sets the tag for connection statistics.
         *
         * @param val The tag to set.
         * @return This builder instance for chaining.
         */
        public Builder connectStatsTag(int val) {
            connectStatsTag = val;
            return this;
        }

        /**
         * Sets the tag for read operation statistics.
         *
         * @param val The tag to set.
         * @return This builder instance for chaining.
         */
        public Builder readStatsTag(int val) {
            readStatsTag = val;
            return this;
        }

        /**
         * Sets the tag for write operation statistics.
         *
         * @param val The tag to set.
         * @return This builder instance for chaining.
         */
        public Builder writeStatsTag(int val) {
            writeStatsTag = val;
            return this;
        }

        /**
         * Sets the number of retry attempts for tasks.
         *
         * @param val The number of attempts to set.
         * @return This builder instance for chaining.
         * @throws IllegalArgumentException If the value is negative.
         */
        public Builder taskRetryTimes(int val) {
            if (val < 0) {
                throw new IllegalArgumentException("Task retry times cannot be negative.");
            }
            taskRetryTimes = val;
            return this;
        }

        /**
         * Builds and returns an `Options` instance with the configured settings.
         *
         * @return A new `Options` instance.
         * @throws IllegalArgumentException If any required parameter is invalid or missing.
         */
        @NonNull
        public Options build() {
            if (headParserProvider == null) {
                throw new IllegalArgumentException("Head parser provider has not been set.");
            }
            if (addressList == null && clientInitializerProvider == null) {
                throw new IllegalArgumentException("address list or client initializer" +
                        " provider should be set.");
            }
            if (socketFactoryProvider == null) {
                socketFactoryProvider = new DefaultSocketFactoryProvider();
            }
            if (callbackExecutor == null) {
                callbackExecutor = Executors.defaultMainThreadExecutor();
            }
            if (connectExecutor == null) {
                connectExecutor = Executors.defaultConnectExecutor();
            }
            if (codecExecutor == null) {
                codecExecutor = Executors.defaultCodecExecutor();
            }
            if (writeExecutor == null) {
                writeExecutor = Executors.defaultWriteExecutor();
            }
            return new Options(this);
        }

        @NonNull
        public SocketClient open() {
            return EasySocket.getInstance().open(build());
        }
    }
}

File: lib/src/commonMain/kotlin/com/orientsec/easysocket/Options.kt
package com.orientsec.easysocket

import com.orientsec.easysocket.client.ClientInitializer
import com.orientsec.easysocket.push.PushManager
import com.orientsec.easysocket.request.Decoder
import com.orientsec.easysocket.request.Request
import com.orientsec.easysocket.session.SessionFactory
import com.orientsec.easysocket.session.SessionInitializer
import com.orientsec.easysocket.session.ktor.KtorSessionFactory
import com.orientsec.easysocket.utils.NoTrafficProfiler
import com.orientsec.easysocket.utils.Platform
import com.orientsec.easysocket.utils.TrafficProfiler
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers

/**
 * EasySocket 库的配置选项类。
 *
 * 包含 Socket 连接的各种设置，如连接超时、心跳配置、重连策略、
 * 编解码调度器、回调调度器等。通过 [Builder] 构建，支持 DSL 风格配置。
 *
 * 使用示例：
 * ```kotlin
 * val options = Options.build {
 *     name = "my-socket"
 *     isDebuggable = true
 *     useAndroidDefaults()
 *     headParserProvider = { MyHeadParser() }
 *     addressList = listOf(Address("192.168.1.1", 8080))
 *     pulseIntervalSeconds = 60
 *     reconnectPolicy = ReconnectPolicy.ALWAYS
 * }
 * ```
 */
class Options private constructor(
    /** 连接名称，用于日志标识 */
    val name: String,
    /** 是否启用调试模式，调试模式下会输出详细日志 */
    val isDebuggable: Boolean,
    /** 最低日志输出级别，低于此级别的日志将被忽略 */
    val minLogLevel: Int,
    /** 心跳请求提供者，返回心跳请求对象 */
    val pulseRequestProvider: Provider<Request<Boolean>>?,
    /** 心跳响应解码器提供者，用于解码心跳响应 */
    val pulseDecoderProvider: Provider<Decoder<Boolean>>?,
    /** 协议头解析器提供者，必须设置 */
    val headParserProvider: Provider<HeadParser>,
    /** 推送消息管理器提供者，可选 */
    val pushManagerProvider: Provider<PushManager<*, *>>?,
    /** 客户端初始化器提供者，用于动态获取服务器地址列表 */
    val clientInitializerProvider: Provider<ClientInitializer>?,
    /** 会话初始化器提供者，用于连接成功后的登录等操作 */
    val sessionInitializerProvider: Provider<SessionInitializer>?,
    /** 会话工厂，负责创建具体的连接会话 */
    val sessionFactory: SessionFactory,
    /** 流量统计器，用于标记 Socket 流量 */
    val trafficProfiler: TrafficProfiler,
    /** 编解码调度器，用于执行数据的编码和解码操作 */
    val codecDispatcher: CoroutineDispatcher,
    /** 回调调度器，用于执行用户回调，默认为主线程 */
    val callbackDispatcher: CoroutineDispatcher,
    /** 服务器地址列表，与 [clientInitializerProvider] 二选一 */
    val addressList: List<Address>?,
    /** 单次读取最大数据量（KB），防止读取超大包导致 OOM */
    val maxReadSizeKb: Int,
    /** 请求超时时间（毫秒），发送成功后等待响应的最大时间 */
    val requestTimeoutMills: Int,
    /** 连接超时时间（毫秒） */
    val connectTimeoutMills: Int,
    /** Tls超时时间（毫秒） */
    val tlsTimeoutMills: Int,
    /** 心跳间隔时间（秒），最小 30 秒 */
    val pulseIntervalSeconds: Int,
    /** 心跳最大丢失次数，超过后认为连接断开 */
    val pulseMaxLostTimes: Int,
    /** 后台活跃持续时间（秒），应用进入后台后保持连接的时间 */
    val backgroundActiveDurationSeconds: Int,
    /** 重连策略，控制是否及何时自动重连 */
    val reconnectPolicy: ReconnectPolicy,
    /** 每个地址的重试次数，超过后切换到下一个地址 */
    val retryTimesPerAddress: Int,
    /** 重连间隔时间（毫秒），两次连接尝试之间的最小间隔 */
    val connectIntervalMillis: Int,
    /** 连接操作的流量统计标签 */
    val connectStatsTag: Int,
    /** 读取操作的流量统计标签 */
    val readStatsTag: Int,
    /** 写入操作的流量统计标签 */
    val writeStatsTag: Int,
    /** 任务重试次数，连接不可用时任务的最大重试次数 */
    val taskRetryTimes: Int
) {

    companion object {
        /**
         * DSL 函数，用于构建 [Options] 实例。
         *
         * @param block 配置 lambda
         * @return 构建完成的 [Options] 实例
         */
        inline fun build(block: Builder.() -> Unit): Options = Builder().apply(block).build()
    }

    /**
     * Options 的构建器，提供 DSL 风格的配置方式。
     * 所有配置项都有合理的默认值，只需设置必要的项即可。
     */
    class Builder {
        /** 连接名称，用于日志标识 */
        var name: String = ""

        /** 是否启用调试模式 */
        var isDebuggable: Boolean = false

        /** 最低日志输出级别，默认 INFO */
        var minLogLevel: Int = Platform.LogLevel.INFO
            set(value) {
                require(value >= 0) { "Minimum log level must be non-negative." }
                field = value
            }

        /** 心跳请求提供者 */
        var pulseRequestProvider: Provider<Request<Boolean>>? = null

        /** 心跳响应解码器提供者 */
        var pulseDecoderProvider: Provider<Decoder<Boolean>>? = null

        /** 协议头解析器提供者，必须设置 */
        var headParserProvider: Provider<HeadParser>? = null

        /** 推送消息管理器提供者 */
        var pushManagerProvider: Provider<PushManager<*, *>>? = null

        /** 客户端初始化器提供者 */
        var clientInitializerProvider: Provider<ClientInitializer>? = null

        /** 会话初始化器提供者 */
        var sessionInitializerProvider: Provider<SessionInitializer>? = null

        /** 会话工厂，必须设置 */
        var sessionFactory: SessionFactory? = null

        /** 流量统计器，默认为空实现 */
        var trafficProfiler: TrafficProfiler = NoTrafficProfiler

        /** 编解码调度器，默认使用 Dispatchers.Default */
        var codecDispatcher: CoroutineDispatcher = Dispatchers.Default

        /** 回调调度器，默认使用平台主线程调度器 */
        var callbackDispatcher: CoroutineDispatcher = Platform.mainDispatcher

        /** 服务器地址列表 */
        var addressList: List<Address>? = null
            set(value) {
                require(value == null || value.isNotEmpty()) { "Address list cannot be empty." }
                field = value
            }

        /** 单次读取最大数据量（KB），默认 1024KB */
        var maxReadSizeKb: Int = 1024
            set(value) {
                require(value > 0) { "Max read data size in KB must be positive." }
                field = value
            }

        /** 请求超时时间（毫秒），默认 5000ms */
        var requestTimeoutMills: Int = 5000
            set(value) {
                require(value > 0) { "Request time out must be positive." }
                field = value
            }

        /** 连接超时时间（毫秒），默认 5000ms */
        var connectTimeoutMills: Int = 5000
            set(value) {
                require(value > 0) { "Connect time out must be positive." }
                field = value
            }

        /** 连接超时时间（毫秒），默认 5000ms */
        var tlsTimeoutMillis: Int = 5000
            set(value) {
                require(value > 0) { "TLS time out must be positive." }
                field = value
            }

        /** 心跳间隔时间（秒），默认 60 秒，最小 30 秒 */
        var pulseIntervalSeconds: Int = 60
            set(value) {
                require(value >= 30) { "Pulse rate must be at least 30 seconds." }
                field = value
            }

        /** 心跳最大丢失次数，默认 2 次 */
        var pulseMaxLostTimes: Int = 2
            set(value) {
                require(value >= 0) { "Pulse lost times cannot be negative." }
                field = value
            }

        /** 后台活跃持续时间（秒），默认 30 秒 */
        var backgroundActiveDurationSeconds: Int = 30
            set(value) {
                require(value >= 0) { "Live time must be positive." }
                field = value
            }

        /** 重连策略，默认 ACTIVE（仅前台重连） */
        var reconnectPolicy: ReconnectPolicy = ReconnectPolicy.ACTIVE

        /** 每个地址的重试次数，默认 0（不重试直接切换） */
        var retryTimesPerAddress: Int = 0
            set(value) {
                require(value >= 0) { "Retry times cannot be negative." }
                field = value
            }

        /** 重连间隔时间（毫秒），默认 3000ms，最小 1000ms */
        var connectIntervalMillis: Int = 3000
            set(value) {
                require(value > 1000) { "Connect interval must be greater than 1000 milliseconds." }
                field = value
            }

        /** 连接操作的流量统计标签 */
        var connectStatsTag: Int = 0x1001

        /** 读取操作的流量统计标签 */
        var readStatsTag: Int = 0x1002

        /** 写入操作的流量统计标签 */
        var writeStatsTag: Int = 0x1003

        /** 任务重试次数，默认 2 次 */
        var taskRetryTimes: Int = 2
            set(value) {
                require(value >= 0) { "Task retry times cannot be negative." }
                field = value
            }

        /**
         * 构建并返回 [Options] 实例。
         *
         * @throws IllegalArgumentException 如果缺少必要配置（headParserProvider、sessionFactory、addressList 或 clientInitializerProvider）
         */
        fun build(): Options {
            val headParser = headParserProvider
                ?: throw IllegalArgumentException("Head parser provider has not been set.")
            if (addressList == null && clientInitializerProvider == null) {
                throw IllegalArgumentException(
                    "address list or client initializer provider " +
                            "should be set."
                )
            }

            return Options(
                name = name,
                isDebuggable = isDebuggable,
                minLogLevel = minLogLevel,
                pulseRequestProvider = pulseRequestProvider,
                pulseDecoderProvider = pulseDecoderProvider,
                headParserProvider = headParser,
                pushManagerProvider = pushManagerProvider,
                clientInitializerProvider = clientInitializerProvider,
                sessionInitializerProvider = sessionInitializerProvider,
                sessionFactory = sessionFactory
                    ?: throw IllegalArgumentException("Session factory must be set."),
                trafficProfiler = trafficProfiler,
                codecDispatcher = codecDispatcher,
                callbackDispatcher = callbackDispatcher,
                addressList = addressList,
                maxReadSizeKb = maxReadSizeKb,
                requestTimeoutMills = requestTimeoutMills,
                connectTimeoutMills = connectTimeoutMills,
                tlsTimeoutMills = tlsTimeoutMillis,
                pulseIntervalSeconds = pulseIntervalSeconds,
                pulseMaxLostTimes = pulseMaxLostTimes,
                backgroundActiveDurationSeconds = backgroundActiveDurationSeconds,
                reconnectPolicy = reconnectPolicy,
                retryTimesPerAddress = retryTimesPerAddress,
                connectIntervalMillis = connectIntervalMillis,
                connectStatsTag = connectStatsTag,
                readStatsTag = readStatsTag,
                writeStatsTag = writeStatsTag,
                taskRetryTimes = taskRetryTimes
            )
        }

        /**
         * 配置使用 Ktor 作为底层网络引擎。
         * 自动设置 [KtorSessionFactory] 作为会话工厂。
         *
         * @return 当前 Builder 实例，支持链式调用
         */
        fun useKtorDefaults(): Builder {
            this.sessionFactory = KtorSessionFactory()
            return this
        }

        /**
         * 构建 Options 并直接打开一个新的 [SocketClient]。
         *
         * @return 新创建的 [SocketClient] 实例
         */
        fun open(): SocketClient {
            return EasySocket.open(build())
        }
    }
}

/**
 * 提供者接口，根据 [SocketClient] 实例创建指定类型的对象。
 * 用于延迟创建依赖 SocketClient 的组件（如 HeadParser、PushManager 等）。
 *
 * @param T 提供的对象类型
 */
typealias Provider<T> = (SocketClient) -> T